docs/notebooks/recovering_from_errors.pct.py

```diff
@@ -0,0 +1,144 @@
+import random
+
+np.random.seed(1793)
+tf.random.set_seed(1793)
+random.seed(0)
+
+# %% [markdown]
+# Sometimes the Bayesian optimization process encounters an error from which we can recover, without the need to restart the run from the beginning. In this tutorial, we'll simulate such an error and show how to recover from it.
+#
+# We'll use a similar setup to the [EI notebook](expected_improvement.ipynb), but use an observer that intermittently breaks when evaluated, and needs manual attention to get running again. We can simulate fixing the observer with its `manual_fix` method.
+
+# %%
+import trieste
+from trieste.acquisition.rule import OBJECTIVE
+from trieste.utils.objectives import branin
+
+
+class FaultyBranin:
+    def __init__(self):
+        self._is_broken = False
+
+    def manual_fix(self):
+        self._is_broken = False
+
+    def __call__(self, x):
+        if self._is_broken:
+            raise Exception("Observer is broken")
+
+        if random.random() < 0.3:
+            self._is_broken = True
+
+        return {OBJECTIVE: trieste.data.Dataset(x, branin(x))}
+
+
+observer = FaultyBranin()
+
+# %% [markdown]
+# ## Set up the problem
+# We'll use the same set up as before, except for the acquisition rule, where we'll use `TrustRegion`, which will better illustrate how to recover.
+
+# %%
+import gpflow
+
+search_space = trieste.space.Box(
+    tf.cast([0.0, 0.0], tf.float64), tf.cast([1.0, 1.0], tf.float64)
+)
+initial_data = observer(search_space.sample(5))
+
+variance = tf.math.reduce_variance(initial_data[OBJECTIVE].observations)
+kernel = gpflow.kernels.Matern52(variance, [0.2, 0.2]) + gpflow.kernels.White(1e-12)
+gpr = gpflow.models.GPR(
+    initial_data[OBJECTIVE].astuple(), kernel, noise_variance=1e-5
+)
+gpflow.set_trainable(gpr.likelihood, False)
+models = {OBJECTIVE: trieste.models.GaussianProcessRegression(gpr)}
+
+acquisition_rule = trieste.acquisition.rule.TrustRegion()
+
+# %% [markdown]
+# ## Run the optimization loop
+#
+# The optimizer returns an `OptimizationResult`. We can use that object's `astuple` method to get a 2-tuple of the final result and the history of the successful optimization steps.
+
+# %%
+bo = trieste.bayesian_optimizer.BayesianOptimizer(observer, search_space)
+
+result, history = bo.optimize(15, initial_data, models, acquisition_rule).astuple()
+
+# %% [markdown]
+# We can see from the logs that the optimization loop failed, and this can be sufficient to know what to do next if we're working in a notebook. However, sometimes our setup means we don't have access to the logs. We'll pretend from here that's the case.
+
+# %% [markdown]
+# ## Handling success
+#
+# We don't know if the optimization completed successfully or not, so we'll only try to access and plot the data if it was successful. We can find out if this was the case with the `is_ok` attribute on the `Result`. If it was successful, we know there is data in the `result`, which we can `unwrap` and view.
```

Comment:
It's a bit confusing that first you only talk about `OptimizationResult`, and then that _contains_ the `Result` that you now talk about here (which from reading the code I understand to be the first element of .astuple()). What might be a way of explaining that difference more clearly?

Reply:
i've extended the markdown section